### PR TITLE
Update test base image and move CI to Github workflow

### PR DESCRIPTION
**Pull request by 3nids**
Dependency image has been moved to ubuntu focal (this means Qt 5.12)

CI is moved away from Travis to Github workflow
Caching is working (ccache)

former PR #39815

### CODE REVIEW

**Comment by m-kuhn** [comment]
Nice!
Was renaming the "stale" workflows done on purpose?

**Comment by 3nids** [comment]
> Was renaming the "stale" workflows done on purpose?

other tests are disabled for now to avoid wasting resources

**Comment by 3nids** [comment]
currently failing tests:

ProcessingQgisAlgorithmsTestPt1
ProcessingQgisAlgorithmsTestPt2
ProcessingQgisAlgorithmsTestPt3
ProcessingQgisAlgorithmsTestPt4
ProcessingGdalAlgorithmsRasterTest
ProcessingGdalAlgorithmsVectorTest
ProcessingGrass7AlgorithmsImageryTest
ProcessingGrass7AlgorithmsRasterTest
ProcessingGrass7AlgorithmsVectorTest
ProcessingSagaAlgorithmsTest
qgis_arcgisrestutilstest
qgis_authmanagertest
qgis_coordinatereferencesystemtest
qgis_geometrytest
qgis_geonodeconnectiontest
qgis_maprendererjobtest
qgis_networkaccessmanagertest
qgis_painteffecttest
qgis_pallabelingtest
qgis_ogrproviderguitest
qgis_3drenderingtest
qgis_processingtest
qgis_geometrycheckstest
PyQgsActionManager
PyQgsPythonProvider
PyQgsAnnotation
PyQgsCoordinateOperationWidget
PyQgsGeometryTest
PyQgsMapLayerComboBox
PyQgsMapLayerProxyModel
PyQgsOGRProviderGpkg
PyQgsProcessExecutable
PyQgsQueryResultModel
PyQgsShapefileProvider
PyQgsSvgCache
PyQgsSpatialiteProvider
PyQgsVectorFileWriter
PyQgsVectorLayer
PyQgsVirtualLayerProvider
PyQgsDBManagerGpkg
PyQgsFileDownloader
PyQgsRelationEditWidget
PyQgsAuthManagerPasswordPostgresTest
PyQgsProviderConnectionPostgres
PyQgsOracleProvider
PyQgsServerWMS
PyQgsServerLandingPage
PyQgsServerWMSGetMap
PyQgsServerWMSGetFeatureInfo
PyQgsServerWMSGetPrint
PyQgsServerWMSGetPrintExtra
PyQgsServerWMSDimension
PyQgsAuthManagerPKIOWSTest
PyQgsAuthManagerOAuth2OWSTest

**Comment by elpaso** [comment]
@3nids I can take care of the following tests, but it will take some time I guess:


PyQgsFileDownloader
PyQgsAuthManagerPasswordPostgresTest
PyQgsProviderConnectionPostgres
PyQgsServerWMS
PyQgsServerLandingPage
PyQgsServerWMSGetMap
PyQgsServerWMSGetFeatureInfo
PyQgsServerWMSGetPrint
PyQgsServerWMSGetPrintExtra
PyQgsServerWMSDimension
PyQgsAuthManagerPKIOWSTest
PyQgsAuthManagerOAuth2OWSTest

**Comment by 3nids** [comment]
@elpaso I advice to run this locally with deps Docker image as is done in the workflow https://github.com/qgis/QGIS/blob/test-focal/.github/workflows/run-tests.yml

**Comment by elpaso** [comment]
@3nids I'm testing one by one and I've started with `PyQgsPythonProvider` but I cannot make it fail locally, I replicated the docker environment on my machine (I've done it before when a test was failing on Travis so it's not new stuff for me) and I'm testing with  `docker-compose -f .docker/docker-compose-testing.yml run qgis-deps /root/QGIS/.docker/docker-qgis-test.sh` (after adding `-R PyQgsPythonProvider` to ctest command line to execute only this one).

The test passes, and it also passes if I `docker exec -it` and run it manually from inside the docker container.

Full output:

```
$ docker-compose  -p QGIS_workflow_testing  -f .docker/docker-compose-testing.yml run qgis-deps /root/QGIS/.docker/docker-qgis-test.sh
Creating qgis_workflow_testing_qgis-deps_run ... done
Load Postgres database...🐘
Wait a moment while loading PostGreSQL database.
 done 🥩
Restoring postgres test data ...
Restoring tests/testdata/provider/testdata_pg.sql
Restoring tests/testdata/provider/testdata_pg_relations.sql
Restoring tests/testdata/provider/testdata_pg_reltests.sql
Restoring tests/testdata/provider/testdata_pg_role.sql
Restoring tests/testdata/provider/testdata_pg_vectorjoin.sql
Restoring tests/testdata/provider/testdata_pg_hstore.sql
Restoring tests/testdata/provider/testdata_pg_array.sql
Restoring tests/testdata/provider/testdata_pg_raster.sql
Restoring tests/testdata/provider/testdata_pg_topology.sql
Restoring tests/testdata/provider/testdata_pg_domain.sql
Restoring tests/testdata/provider/testdata_pg_json.sql
Restoring tests/testdata/provider/testdata_pg_pointcloud.sql
Restoring tests/testdata/provider/testdata_pg_bigint_pk.sql
Restoring tests/testdata/provider/testdata_pg_hasspatialindex.sql
Database qgis_test populated and ready for use with 'service=qgis_test'
Postgres test data restored ...
Flaky tests are skipped!
List of skipped tests: PyQgsLocalServer|PyQgsPalLabelingServer|qgis_composerutils|PyQgsAppStartup|qgis_spelling|qgis_sipify|qgis_sip_include|qgis_sip_uptodate|PyQgsAuthManagerPKIPostgresTest|PyQgsAuthManagerOgrPostgresTest|PyQgsDbManagerPostgis|qgis_openclutilstest|qgis_layerdefinition|PyQgsProviderConnectionMssql|qgis_filedownloader|qgis_wcsprovidertest|PyQgsWFSProviderGUI|qgis_ziplayertest|PyQgsMssqlProvider|PyQgsPostgresRasterProvider
travis_fold:start:build
Running tests...
Run dashboard with model Experimental
   Source directory: /root/QGIS
   Build directory: /root/QGIS/build
   Reading ctest configuration file: /root/QGIS/CTestConfig.cmake
   Site: github.com
   Build name: (empty)
   Use Experimental tag: 20210130-0848
Build project
   Each symbol represents 1024 bytes of output.
   '!' represents an error and '*' a warning.
    ........................ Size of output: 24K
   0 Compiler errors
   0 Compiler warnings
Test project /root/QGIS/build
    Start 322: PyQgsPythonProvider
1/1 Test #322: PyQgsPythonProvider ..............   Passed    8.69 sec

travis_fold:end:build
100% tests passed, 0 tests failed out of 1

Total Test time (real) =   8.71 sec
```

Am I missing something?

**Comment by elpaso** [comment]
@3nids the tests I've assigned to myself are all passing after https://github.com/qgis/QGIS/pull/41260/commits/3763fc56a9867a59f0fff035492b24bdcb4818bb on my local docker environment.

Let's see how the workflow goes. I just hope we are not switching from bad Travis to worse GitHub... 

If I sum up all the time that I've wasted on Travis nonsense failures during the last years I could have rewritten QGIS in `nodejs` :smile_cat: .

**Comment by elpaso** [comment]
@3nids nevermind, I had some deps image leftover, after rebuilding the image I can reproduce most of the issues.

**Comment by nyalldawson** [comment]
@3nids where's the appropriate place to set XDG_RUNTIME_DIR ? I think if we correctly set that at /tmp we may resolve a few more issues...

**Comment by nyalldawson** [comment]
Looks like some of the failures are coming as a result of using the c locale, which means that QString::localeAwareCompare is case sensitive, not insensitive as the tests expect. Can we hardcode a locale of EN-US on the image?

**Comment by elpaso** [comment]
You can try to put locale in the docker env file, it is read by the compose

**Comment by nyalldawson** [comment]
@3nids I think the push to cdash changes you made aren't working properly -- we're no longer getting the results pushed to dash

**Comment by 3nids** [comment]
failing test for 3D rendering shows quite different control images: https://cdash.orfeo-toolbox.org/testDetails.php?test=26523817&build=58951
What shall we do with these?

**Comment by 3nids** [comment]
updated lists of remaining tests

ProcessingGdalAlgorithmsRasterTest
ProcessingGdalAlgorithmsVectorTest
ProcessingGrass7AlgorithmsImageryTest
ProcessingGrass7AlgorithmsRasterTest
ProcessingGrass7AlgorithmsVectorTest
ProcessingQgisAlgorithmsTestPt1
ProcessingQgisAlgorithmsTestPt2
ProcessingQgisAlgorithmsTestPt3
ProcessingQgisAlgorithmsTestPt4
ProcessingSagaAlgorithmsTest
PyQgsCoordinateOperationWidget
PyQgsGeometryTest
PyQgsGoogleMapsGeocoder
PyQgsOracleProvider
PyQgsRelationEditWidget
PyQgsServerLandingPage
PyQgsServerWMSDimension
PyQgsServerWMSGetMap
qgis_3drenderingtest
qgis_coordinatereferencesystemtest
qgis_geometrycheckstest
qgis_geometrytest
qgis_pallabelingtest

+ Oracle has bee disabled

**Comment by nyalldawson** [comment]
Rebased to get fix for flaky tile manager test

**Comment by nyalldawson** [comment]
@3nids the output from test_qgscoordinatereferencesystem.cpp is always truncated (even in the raw github action logs) and I can't see what's actually failing there. Do you know why this is occurring?

**Comment by 3nids** [comment]
> the output from test_qgscoordinatereferencesystem.cpp is always truncated (even in the raw github action logs) and I can't see what's actually failing there. Do you know why this is occurring?

I had a look and I don't see any reason why the log would be truncated. We can try not to use the ctest2.ci python script, to see if this helps?

**Comment by elpaso** [comment]
PyQgsServerWMSGetMap is passing in my local docker.

```
1/1 Test #686: PyQgsServerWMSGetMap .............   Passed    9.67 sec

::endgroup::
100% tests passed, 0 tests failed out of 1
```

**Comment by nyalldawson** [comment]
Ok, I'm disabling the last two failing tests:

- PyQgsOracleProvider: as noted above we can't get the Oracle container running under Github Actions. @troopa81  this may be something you are able to fix? (as your team is most knowledgeable about this container setup). For now we'll just proceed without CI testing of Oracle.
- PyQgsRelationEditWidget: causes a crash for unknown reasons. This warrants further investigation, but I don't think we should hold back this work further because of this one test. Basically, we are running MANY more tests on GA vs Travis, so the net win far outweighs one new blocklisted test.

@3nids I think we can merge now... (and then deal with the nightmare of backporting this... :scream: )

**Comment by nyalldawson** [comment]
@3nids all tests are passing, but now we've got a new error on the test runners stage. Do you know what this is about?

**Comment by roya0045** [comment]
> @3nids all tests are passing, but now we've got a new error on the test runners stage. Do you know what this is about?

https://stackoverflow.com/questions/43099116/error-the-input-device-is-not-a-tty docker parameter

**Comment by nyalldawson** [comment]
PyQgsRelationEditWidget -- actually I've fixed the cause of that crash now, so I've re-enabled the test

File: .github/workflows/run-tests.yml [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+      - name: Push image
+        if: github.event_name == 'pull_request' && contains( github.event.pull_request.labels.*.name, 'Push Docker Image')
+        env:
+          GITHUB_PR_NUMBER: ${{github.event.number}}
+        run: |
+          docker login -u="${DOCKER_USERNAME}" -p="${DOCKER_PASSWORD}"
+          docker commit qgis_container qgis/qgis-pr-build:PR${GITHUB_PR_NUMBER}
+          docker push qgis/qgis-pr-build:PR${GITHUB_PR_NUMBER}
+
+      - name: run unit tests
```
Comment (nirvn) [inline .github/workflows/run-tests.yml:128]: I'd capitalize 'run' so its consistent with the other section titles:
image

**Comment by troopa81** [comment]
> PyQgsOracleProvider: as noted above we can't get the Oracle container running under Github Actions. @troopa81 this may be something you are able to fix? (as your team is most knowledgeable about this container setup). For now we'll just proceed without CI testing of Oracle.

I agree. I'll take a look into this and will restore the Oracle tests later. Thank you all for this outstanding work.

**Comment by nirvn** [comment]
:tada:

**Comment by lbartoletti** [comment]
Thanks!

File: CMakeLists.txt [hunk 2]
```diff
@@ -532,18 +532,19 @@ set(QML_IMPORT_PATH "${CMAKE_RUNTIME_OUTPUT_DIRECTORY}" CACHE PATH "QML director
...
-add_custom_target(check COMMAND xvfb-run --server-args=-screen\ 0\ 1024x768x24 ctest --output-on-failure)
+  set (PUSH_TO_CDASH FALSE CACHE BOOL "Determines whether test results should be pushed to CDASH site")
+  set(QT_USE_QTTEST TRUE)
+  enable_testing()
+  # Adds some testing specific build targets e.g. make Experimental
+  include(Dart)
+  # Additional test configuration options e.g. max upload size of test report
+  configure_file(
+    "${CMAKE_SOURCE_DIR}/cmake_templates/CTestCustom.cmake.in"
+    "${CMAKE_BINARY_DIR}/CTestCustom.cmake"
```
Comment (jef-n) [inline CMakeLists.txt:543]: @3nids hm, does this break submission of the nightlies?  Probably along with something else. For some reason ``CTestCustom.cmake`` is apparently not picked up anymore and submission is tried to a blank ``https://``.  Copying ``CTestCustom.cmake`` back to ``${CMAKE_SOURCE_DIR}`` helps at some places, but not everywhere (or on all branches).  Any hints?

Comment (3nids) [reply CMakeLists.txt:543]: Submission to the dashboard? Have you defined CTEST_BUILD_NAME?

Comment (jef-n) [reply CMakeLists.txt:543]: Yes, to the dashboard.   The nightlies set `BUILDNAME` when configuring cmake.

Windows:

https://github.com/qgis/QGIS/blob/570efee7255d9201c460da095421eee2ea4882aa/ms-windows/osgeo4w/package-nightly.cmd#L120

Debian/Ubuntu:

https://github.com/qgis/QGIS/blob/570efee7255d9201c460da095421eee2ea4882aa/debian/rules#L84

And the osgeo4w testing builds work alike:

https://github.com/jef-n/OSGeo4W/blob/16ae5f88c12fe5678ca0c4c7efd2c352b0e20a91/src/qgis-dev/osgeo4w/package.sh#L131

But the results vary between platforms, branches and repos.

Comment (jef-n) [reply CMakeLists.txt:543]: It started: https://cdash.orfeo-toolbox.org/index.php?project=QGIS&date=2021-02-20

**Comment by 3nids** [comment]
Adding `-DPUSH_TO_CDASH=true` will push to CDASH.
For the the build name, it now relies on an env var (`CTEST_BUILD_NAME`). I cannot how it used to work with `BUILDNAME`. We could switch this to a cmake var, that would be cleaner?

**Comment by 3nids** [comment]
is actually the `BUILDNAME` automatically picked up by CTEST?

**Comment by jef-n** [comment]
PUSH_TO_CDASH is set.

Not sure how all this works together anymore.  I think it has worked this way since the nightlies were introduced - to run the tests (HF lisboa I think).

BUILDNAME is only referenced in .ci/config.ctest and qgis-test.ctest.example.

ctest(1) says BUILDNAME is the CTest module variable and CTEST_BUILD_NAME the CTest script variable for the BuildName setting for the "CTest submit step".   But I don't see why the buildname would reset the SubmitURL either - as the CTEST_SUBMIT_URL points at the dashboard in CMakeCache.txt and still it tries to upload to "http://".

And the nightlies run the cmake targets and not ctest directly.

**Comment by 3nids** [comment]
I changed CTestConfig.cmake to be a template file now (.in), which is configured only when PUSH_TO_CDASH is set.
Do you have CTestConfig.cmake in the build dir?

**Comment by jef-n** [comment]
Yes.   ouch, I was talking about CTestConfig.cmake all along - but I linked the wrong line (a few lines above PUSH_TO_CDASH part instead of this within).  6e5f4f36166c5a29759f4f78abeab6e9b579e2aa copies CTestConfig.cmake.

**Comment by 3nids** [comment]
cool, sorry for the troubles!

**Comment by jef-n** [comment]
> cool, sorry for the troubles!

unfortunately not cool - as it doesn't help.  It looked like it did locally, but it doesn't help on build machines.   On all branches the final SubmitURL is http:// now - even on master with 6e5f4f3.   

Except for release-3_16 on windows  64bit - where it fails to build (because somebody apparently reordered the include directories and PROJ_INCLUDE_DIR now comes too late to override proj include of the release with proj nightly's includes; but that's a different story), but it is submitted.  It also fails on 32bit, but that's not submitted.

release-3_18 succeeds to build, but also fails to submit.

but master is not built there anymore (because "current" osgeo4w doesn't fulfil the dependencies since the aggressive dependency lift; another different story).  For the "new" osgeo4w I didn't setup nightlies yet (or better put setup workflows and watched it run out of diskspace…; yet another different story).

The same on the qgis2 with the debian/ubuntu builds - `http://` - but without any exceptions (but the last "nightly" round took 3d10h53m14s).

In the previous round of windows builds it was release-3_18 on 64bit that was submitted - one before I fixed an include that broke the build and one after. Odd.

Maybe somehow related to the cmake build tag?   Seems to switch at different points.   Does any of that ring a bell?

**Comment by jef-n** [comment]
confused? -v ;)

**Comment by roya0045** [comment]
> confused? -v ;)

It was the closest to a sadface. Would it even be possible to have  some tests to prevent some compound issues? Though on the other hand I doubt such drastic changes should repeat soon enough..

**Comment by 3nids** [comment]
cmake_templates/CTestConfig.cmake.in is only configured by the main CMakeLists.txt when ENABLE_TESTS is true.
https://github.com/qgis/QGIS/blob/master/CMakeLists.txt#L536-L551
Actually PUSH_TO_CDASH is only defined when tests are enabled. Is this the issue?

Otherwise, have you set CTEST_BUILD_NAME env var or just BUILDNAME cmake var?

**Pull request merged**